packages/service-library/src/servicelib/background_task.py

```diff
@@ -56,21 +58,43 @@ async def start_periodic_task(
         )
 
 
+async def cancel_task(
+    task: asyncio.Task,
+    *,
+    timeout: float | None,
+    cancellation_attempts: int = _MAX_TASK_CANCELLATION_ATTEMPTS,
+) -> None:
+    """Reliable task cancellation. Some libraries will just hang without
+    cancelling the task. It is important to retry the operation to provide
+    a timeout in that situation to avoid forever pending tasks.
+
+    :param task: task to be canceled
+    :param timeout: total duration (in seconds) to wait before giving
+        up the cancellation. If None it waits forever.
+    :raises TryAgain: raised if cannot cancel the task.
+    """
+    async for attempt in AsyncRetrying(
+        stop=stop_after_attempt(cancellation_attempts), reraise=True
+    ):
+        with attempt:
+            task.cancel()
+            _, pending = await asyncio.wait((task,), timeout=timeout)
+            if pending:
+                logger.info(
+                    "tried to cancel '%s' but timed-out! %s", task.get_name(), pending
+                )
+                raise TryAgain()
+
+
 async def stop_periodic_task(
-    asyncio_task: asyncio.Task, *, timeout: Optional[float] = None
+    asyncio_task: asyncio.Task, *, timeout: float | None = None
 ) -> None:
     with log_context(
         logger,
         logging.INFO,
         msg=f"cancel periodic background task '{asyncio_task.get_name()}'",
     ):
-        asyncio_task.cancel()
-        _, pending = await asyncio.wait((asyncio_task,), timeout=timeout)
-        if pending:
-            logger.warning(
-                "periodic background task '%s' did not cancel properly and timed-out!",
-                f"{asyncio_task.get_name()}",
-            )
+        await cancel_task(asyncio_task, timeout=timeout)
```

Comment:
```suggestion
        await _cancel_task(asyncio_task, timeout=timeout)
```

Reply:
I think it's OK to be public. I think it can be used in other places as well.